File: forms/neurodiversity-adjustment-review/front-end-with-svelte/src/lib/engine/wellbeing-rules.ts
import type { NeurodiversityAdjustmentReview, WellbeingRiskBand, FiredRule } from './types';
import { anyNotWorking } from './utils';

/** Band severity order (lowest → highest); the maximum fired band wins. */
const BAND_ORDER: Record<Exclude<WellbeingRiskBand, ''>, number> = {
	ok: 0,
	caution: 1,
	'high-risk': 2
};

/**
 * Axis B — wellbeing risk.
 *
 * Every rule below is evaluated; all that fire are recorded in the audit trail.
 * The final band is the maximum severity among the fired rules. When nothing
 * fires the band is `ok` and R-WELL-OK is emitted.
 *
 * An escalation, declining wellbeing, or a dissatisfied worker each drive the
 * band to high-risk — the principal wellbeing-risk signals of a review.
 */
export function gradeWellbeing(r: NeurodiversityAdjustmentReview): {
	wellbeingRiskBand: WellbeingRiskBand;
	firedRules: FiredRule[];
} {
	const firedRules: FiredRule[] = [];

	if (r.escalated) {
		firedRules.push({
			ruleId: 'R-WELL-ESCALATED',
			axis: 'wellbeing',
			category: 'high-risk',
			description: 'Matter escalated.'
		});
	}

	if (r.wellbeingChange === 'worse') {
		firedRules.push({
			ruleId: 'R-WELL-DECLINED',
			axis: 'wellbeing',
			category: 'high-risk',
			description: "Worker's wellbeing has worsened since the adjustments."
		});
	}

	if (r.workerSatisfied === 'no') {
		firedRules.push({
			ruleId: 'R-WELL-DISSATISFIED',
			axis: 'wellbeing',
			category: 'high-risk',
			description: 'Worker is not satisfied the adjustments meet their needs.'
		});
	}

	if (anyNotWorking(r)) {
		firedRules.push({
			ruleId: 'R-WELL-NOT-WORKING',
			axis: 'wellbeing',
			category: 'caution',
			description: 'At least one adjustment is not working.'
		});
	}

	if (r.workerSatisfied === 'partially') {
		firedRules.push({
			ruleId: 'R-WELL-PARTIAL-SATISFACTION',
			axis: 'wellbeing',
			category: 'caution',
			description: 'Worker is only partially satisfied.'
		});
	}

	if (r.barriersDetail.trim() !== '') {
		firedRules.push({
			ruleId: 'R-WELL-BARRIERS',
			axis: 'wellbeing',
			category: 'caution',
			description: 'Remaining barriers reported.'
		});
	}

	if (firedRules.length === 0) {
		firedRules.push({
			ruleId: 'R-WELL-OK',
			axis: 'wellbeing',
			category: 'ok',
			description: 'No wellbeing risk detected from the review.'
		});
		return { wellbeingRiskBand: 'ok', firedRules };
	}

	let band: Exclude<WellbeingRiskBand, ''> = 'ok';
	for (const rule of firedRules) {
		const c = rule.category as Exclude<WellbeingRiskBand, ''>;
		if (c in BAND_ORDER && BAND_ORDER[c] > BAND_ORDER[band]) band = c;
	}
	return { wellbeingRiskBand: band, firedRules };
}

File: forms/neurodiversity-adjustment-review/back-end-with-loco/src/engine/wellbeing_rules.rs
//! Axis B — wellbeing risk.
//!
//! Risk ladder `ok < caution < high-risk`. Every rule whose condition holds is
//! collected into the audit trail; the band is the highest-ranked one among the
//! rules that fired. If no rule fires, the review is `ok` and `R-WELL-OK` is
//! emitted.
//!
//! An escalation, declining wellbeing, or a dissatisfied worker drives
//! `high-risk`; a not-working adjustment, partial satisfaction, or remaining
//! barriers drive `caution`.

use super::types::{FiredRule, NeurodiversityAdjustmentReview, WellbeingRiskBand};
use super::utils::any_not_working;

fn band_rank(band: &str) -> u8 {
    match band {
        "high-risk" => 2,
        "caution" => 1,
        _ => 0,
    }
}

/// Grade the wellbeing-risk band, returning the band plus every rule that fired.
#[must_use]
pub fn grade_wellbeing(
    r: &NeurodiversityAdjustmentReview,
) -> (WellbeingRiskBand, Vec<FiredRule>) {
    let mut fired_rules: Vec<FiredRule> = Vec::new();
    let mut band = "ok".to_string();

    let mut fire =
        |rule_band: &str, rule_id: &str, category: &str, description: &str| {
            if band_rank(rule_band) > band_rank(&band) {
                band = rule_band.to_string();
            }
            fired_rules.push(FiredRule {
                rule_id: rule_id.to_string(),
                axis: "wellbeing".to_string(),
                category: category.to_string(),
                description: description.to_string(),
            });
        };

    if r.escalated {
        fire(
            "high-risk",
            "R-WELL-ESCALATED",
            "escalated",
            "Matter escalated.",
        );
    }
    if r.wellbeing_change == "worse" {
        fire(
            "high-risk",
            "R-WELL-DECLINED",
            "wellbeing-declined",
            "Worker's wellbeing has worsened since the adjustments.",
        );
    }
    if r.worker_satisfied == "no" {
        fire(
            "high-risk",
            "R-WELL-DISSATISFIED",
            "dissatisfied",
            "Worker is not satisfied the adjustments meet their needs.",
        );
    }
    if any_not_working(r) {
        fire(
            "caution",
            "R-WELL-NOT-WORKING",
            "not-working",
            "At least one adjustment is not working.",
        );
    }
    if r.worker_satisfied == "partially" {
        fire(
            "caution",
            "R-WELL-PARTIAL-SATISFACTION",
            "partial-satisfaction",
            "Worker is only partially satisfied.",
        );
    }
    if !r.barriers_detail.trim().is_empty() {
        fire(
            "caution",
            "R-WELL-BARRIERS",
            "barriers",
            "Remaining barriers reported.",
        );
    }

    if fired_rules.is_empty() {
        fired_rules.push(FiredRule {
            rule_id: "R-WELL-OK".to_string(),
            axis: "wellbeing".to_string(),
            category: "ok".to_string(),
            description: "No wellbeing risk detected from the review.".to_string(),
        });
    }

    (band, fired_rules)
}
